psycopg/psycopgmodule.c

```diff
@@ -112,6 +112,44 @@ psyco_connect(PyObject *self, PyObject *args, PyObject *keywds)
     return conn;
 }
 
+#define psyco_parse_dsn_doc \
+"parse_dsn(dsn) -- Parse database connection string.\n\n"
+
+static PyObject *
+psyco_parse_dsn(PyObject *self, PyObject *args)
+{
+    char *dsn, *err;
+    PQconninfoOption *options = NULL, *o;
+    PyObject *res = NULL, *value;
+
+    if (!PyArg_ParseTuple(args, "s", &dsn)) {
+        return NULL;
+    }
+
+    options = PQconninfoParse(dsn, &err);
+    if (!options) {
+        PyErr_Format(PyExc_RuntimeError, "PQconninfoParse: %s: %s", dsn, err);
+        PQfreemem(err);
+        return NULL;
+    }
+
+    res = PyDict_New();
+    for (o = options; o->keyword != NULL; o++) {
+        if (o->val != NULL) {
+            value = PyString_FromString(o->val);
+            if (value == NULL || PyDict_SetItemString(res, o->keyword, value) != 0) {
```

Comment:
This is terrible, separate the error check from PyString_FromString from the SetItemString!

It is also a reference leak: SetItemString doesn't steal the reference. value must be decref'd.